account/accounts.go

```diff
@@ -477,3 +480,109 @@ func (m *Manager) MigrateKeyStoreDir(oldDir, newDir string, addresses []string)
+	}
+	n := int(cryptoJSON.KDFParams["n"].(float64))
+	p := int(cryptoJSON.KDFParams["p"].(float64))
+
+	gethKey := gethkeystore.Key{
+		Id:              decryptedKey.ID,
+		Address:         gethcommon.Address(decryptedKey.Address),
+		PrivateKey:      decryptedKey.PrivateKey,
+		ExtendedKey:     decryptedKey.ExtendedKey,
+		SubAccountIndex: decryptedKey.SubAccountIndex,
+	}
+
+	return gethkeystore.EncryptKey(&gethKey, newPass, n, p)
+}
+
+func (m *Manager) ReEncryptKeyStoreDir(keyDirPath, oldPass, newPass string) error {
+	rencryptFileAtPath := func(tempKeyDirPath, path string, fileInfo os.FileInfo) error {
+		if fileInfo.IsDir() {
+			return nil
+		}
+
+		rawKeyFile, e := ioutil.ReadFile(path)
+		if e != nil {
+			return fmt.Errorf("invalid account key file: %v", e)
+		}
+
+		reEncryptedKey, e := m.ReEncryptKey(rawKeyFile, oldPass, newPass)
+		if e != nil {
+			return fmt.Errorf("unable to re-encrypt key file: %v, path: %s, name: %s", e, path, fileInfo.Name())
+		}
+
+		tempWritePath := filepath.Join(tempKeyDirPath, fileInfo.Name())
+		e = ioutil.WriteFile(tempWritePath, reEncryptedKey, fileInfo.Mode().Perm())
+		if e != nil {
+			return fmt.Errorf("unable write key file: %v", e)
+		}
+
+		return nil
+	}
+
+	parts := strings.Split(keyDirPath, string(os.PathSeparator))
+	keyDirName := parts[len(parts)-1]
+	keyParent := strings.Join(parts[:len(parts)-1], string(os.PathSeparator))
+
+	// backupKeyDirName used to store existing keys before final write
+	backupKeyDirName := keyDirName + "-backup"
+	// tempKeyDirName used to put re-encrypted keys
+	tempKeyDirName := keyDirName + "-re-encrypted"
+	backupKeyDirPath := filepath.Join(keyParent, backupKeyDirName)
+	tempKeyDirPath := filepath.Join(keyParent, tempKeyDirName)
+
+	// create temp key dir
+	err := os.MkdirAll(tempKeyDirPath, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("mkdirall error: %v, tempKeyDirPath: %s", err, tempKeyDirPath)
+	}
+
+	err = filepath.Walk(keyDirPath, func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			os.RemoveAll(tempKeyDirPath)
+			return fmt.Errorf("walk callback error: %v", err)
+		}
+
+		return rencryptFileAtPath(tempKeyDirPath, path, fileInfo)
+	})
+	if err != nil {
+		os.RemoveAll(tempKeyDirPath)
+		return fmt.Errorf("walk error: %v", err)
+	}
+
+	// move existing keys
+	err = os.Rename(keyDirPath, backupKeyDirPath)
+	if err != nil {
+		os.RemoveAll(tempKeyDirPath)
+		return fmt.Errorf("unable to rename keyDirPath to backupKeyDirPath: %v", err)
+	}
+
+	// move tempKeyDirPath to keyDirPath
+	err = os.Rename(tempKeyDirPath, keyDirPath)
+	if err != nil {
+		// if this happens, then the app is probably bricked, because the keystore won't exist anymore
+		// try to restore from backup
+		_ = os.Rename(backupKeyDirPath, keyDirPath)
+		return fmt.Errorf("unable to rename tempKeyDirPath to keyDirPath: %v", err)
+	}
+
+	// remove temp and backup folders and their contents
+	os.RemoveAll(tempKeyDirPath)
+	os.RemoveAll(backupKeyDirPath)
```

Comment:
Nice cleanup of  toxic cryptographic waste! os.RemoveAll can't error?

Reply:
It can, will catch it but I don't think we should throw that error because the process went through. 
I will retry deleting files.

Reply:
Yupp, i wouldn't throw here either but log.error that the user needs to remove the folder manually while they're present